### PR TITLE
[WIP] EVM transformation

### PR DESCRIPTION
**Pull request by jangko**

### CODE REVIEW

**Comment by jangko** [comment]
this PR is ready. review please.
summary:
* GST +57, only 2 left from Frontier and Homestead test suite.
* discover an important fact related to long standing stack overflow issue: `{.computedGoto.}` pragma can eat a lot of stack space. I will transform the EVM into jump table based in another PR.
* we are able to sync pass DAO, I will tell the exact number later, right now I still syncing.

File: nimbus/vm/computation.nim [hunk 1]
```diff
@@ -142,65 +153,68 @@ proc writeContract*(computation: var BaseComputation, fork: Fork): bool =
     if fork < FkHomestead: computation.output = @[]
     result = false
 
-proc transferBalance(computation: var BaseComputation, opCode: static[Op]): bool =
+proc transferBalance(computation: BaseComputation, opCode: static[Op]) =
   if computation.msg.depth > MaxCallDepth:
-    debug "Stack depth limit reached", depth=computation.msg.depth
-    return false
+    computation.setError(&"Stack depth limit reached depth={computation.msg.depth}")
```
Comment (stefantalpalaru) [inline nimbus/vm/computation.nim]: Why is it necessary to copy the string and where are these errors logged through chronicles?

**Comment by jangko** [comment]
The prefix '&' is not a concat operator, it is a strfmt string interpolation shortcut.
The error msg should be appear at transaction tracer, we can easily log it with chronicles at the end of 'executeOpcodes'.
I will add the logging code later when I fix #244

File: nimbus/vm/computation.nim [hunk 2]
```diff
@@ -142,65 +153,68 @@ proc writeContract*(computation: var BaseComputation, fork: Fork): bool =
...
 
-  result = true
+template continuation*(comp: BaseComputation, body: untyped) =
+  # this is a helper template to implement continuation
+  # passing and convert all recursion into tail call
+  var tmpNext = comp.nextProc
+  comp.nextProc = proc() =
+    body
+    tmpNext()
+
```
Comment (stefantalpalaru) [inline nimbus/vm/computation.nim]: There's no continuation being passed here, is it? Just chaining blocks of code.
By the way, is it intentional that the last added block is executed first?

Comment (jangko) [reply nimbus/vm/computation.nim]: > There's no continuation being passed here, is it? Just chaining blocks of code.

traditional cps passing the continuation via proc param, but it doesn't have to. you can pass it via closure too.

> By the way, is it intentional that the last added block is executed first?

yes, it emulates stack behaviour, last in first out, implicit stack. but still follow the principles of continuation passing.

**Pull request merged**